### PR TITLE
Move mod/search to src/Module/Search/Index

### PR DESCRIPTION
**Pull request by MrPetovan**
Part of $4090

Oof.

So I found out that saved search management was done *both* from `/search` and `/network`, which accepted a `search` query string parameter to redirect to `/search`. Also the saved search aside widget was defined in both places as well.

After the dust had settled, there is now a dedicated module and widget for managing saved search, and no `/network` search nonsense.

~~@nupplaphil Once again, I bumped into properties/methods you deprecated whose suggestion for replacement isn't accessible outside of the relevant class. This time I added an accessor for `App->args` property because I was frustrated not to be able to actually implement the correct way of accessing argument. May I please ask you next time you deprecate a public something in favor of a private something to add a publicly accessible alternative?~~
Edit: Secret is using `BaseModule::getClass($class_name)` to retrieve most class instances.

### CODE REVIEW

File: src/App.php [hunk 1]
```diff
@@ -132,6 +132,16 @@ class App
 	 */
 	private $process;
 
+	/**
+	 * Returns the arguments of the query
+	 *
+	 * @return Arguments
+	 */
+	public function getArgs()
```
Comment (nupplaphil) [inline src/App.php]: We don't need this access if we use `BaseObject::getClass(App\Arguments::class)`

File: src/BaseModule.php [hunk 2]
```diff
@@ -155,4 +155,12 @@ public static function checkFormSecurityTokenForbiddenOnError($typename = '', $f
 			throw new \Friendica\Network\HTTPException\ForbiddenException();
 		}
 	}
+
+	/**
+	 * @return App\Arguments
+	 */
+	protected static function getArgs()
```
Comment (nupplaphil) [inline src/BaseModule.php]: This is the same as `return self::getClass(App\Arguments::class)`

File: src/Module/BaseSearchModule.php [hunk 3]
```diff
@@ -62,7 +63,7 @@ public static function performSearch($prefix = '')
 			$header = L10n::t('Forum Search - %s', $search);
 		}
 
-		$pager = new Pager($a->query_string);
+		$pager = new Pager(self::getArgs()->getQueryString());
```
Comment (nupplaphil) [inline src/Module/BaseSearchModule.php]:
```suggestion
		/** @var App\Arguments */
		$args = self::getClass(App\Arguments::class);
		$pager = new Pager($args->getQueryString());
```

File: src/Module/Search/Index.php [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+		}
+
+		$tag = $tag || Config::get('system', 'only_tag_search');
+
+		// Here is the way permissions work in the search module...
+		// Only public posts can be shown
+		// OR your own posts if you are a logged in member
+		// No items will be shown if the member has a blocked profile wall.
+
+		$pager = new Pager(self::getArgs()->getQueryString());
```
Comment (nupplaphil) [inline src/Module/Search/Index.php]:
```suggestion
		/** @var App\Arguments */
		$args = self::getClass(App\Arguments::class);
		$pager = new Pager($args->getQueryString());
```

File: src/Module/Search/Index.php [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+			if (!empty($itemids)) {
+				$params = ['order' => ['id' => true]];
+				$items = Item::selectForUser(local_user(), [], ['id' => $itemids], $params);
+				$r = Item::inArray($items);
+			} else {
+				$r = [];
+			}
+		} else {
+			Logger::info('Start fulltext search for "' . $search . '"');
```
Comment (nupplaphil) [inline src/Module/Search/Index.php]:
```suggestion
			Logger::info('Start fulltext search.', ['search' => $search]);
```

File: src/Module/Search/Index.php [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+			$title = L10n::t('Items tagged with: %s', $search);
+		} else {
+			$title = L10n::t('Results for: %s', $search);
+		}
+
+		$o .= Renderer::replaceMacros(Renderer::getMarkupTemplate('section_title.tpl'), [
+			'$title' => $title
+		]);
+
+		Logger::info('Start Conversation for "' . $search . '"');
```
Comment (nupplaphil) [inline src/Module/Search/Index.php]:
```suggestion
		Logger::info('Start Conversation.', ['search' => $search]);
```

File: src/Module/Search/Saved.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+use Friendica\BaseModule;
+use Friendica\Core\L10n;
+use Friendica\Database\DBA;
+use Friendica\Util\Strings;
+
+class Saved extends BaseModule
+{
+	public static function rawContent()
+	{
+		$action = self::getArgs()->get(2, 'none');
```
Comment (nupplaphil) [inline src/Module/Search/Saved.php]:
```suggestion
		/** @var App\Arguments */
		$args = self::getClass(App\Arguments::class);
		$action = $args->get(2, 'none');
```

File: src/Module/Search/Index.php [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+		// Here is the way permissions work in the search module...
+		// Only public posts can be shown
+		// OR your own posts if you are a logged in member
+		// No items will be shown if the member has a blocked profile wall.
+
+		$pager = new Pager(self::getArgs()->getQueryString());
+
+		if ($tag) {
+			Logger::info('Start tag search for "' . $search . '"');
```
Comment (nupplaphil) [inline src/Module/Search/Index.php]:
```suggestion
			Logger::info('Start tag search.', ['search' => $search]);
```

**Comment by MrPetovan** [comment]
> All deprecated fields or methods in the `App` class are accessible through this method. `self::getClass(...)` will trigger Dice to create this class (or return it of it's cache). So we don't need the App accessor for such entities anymore.

Ooooh, I get it now, it just wasn't clear in the deprecation suggestions.

**Comment by annando** [comment]
It's still not clear to me how to do it with static classes and functions that aren't classes at all (The addons for example).

**Comment by MrPetovan** [comment]
@nupplaphil Is there any counter-indication of making `BaseModule::getClass` public?

**Comment by nupplaphil** [comment]
> 
> 
> @nupplaphil Is there any counter-indication of making `BaseModule::getClass` public so that it can be used in addons?

I found it more clear that a class has to extend `BaseObject` or `BaseModule` to get access to `getClass()` (to hold the usage tight) but there's no other reason :-)

**Comment by nupplaphil** [comment]
> 
> 
> It's still not clear to me how to do it with static classes and functions that aren't classes at all (The addons for example).

You can use static classes as is. Only new classes without static container (like the `Friendica\App\...` classes) can (currently) only be accessed through `self::getClass()`. So we don't create a static container.

There's no difference between:
```php
class ClassA()
{
      public function methodA()
      {
      }
}

class StaticClassA extends BaseObject
{
        public static methodA()
        {
               return self::getClass(ClassA::class)->methodA();
        }
}

StaticClassA::methodA();
```
and
```php
class ClassA()
{
      public function methodA()
      {
      }
}

$classA = BaseObject::getClass(ClassA::class);
$classA->methodA();
```

The main difference is that we don't need a static wrapper for a class.

**Pull request merged**